File: app/src/main/java/com/github/ljarka/bakingapp/MainActivity.java
package com.github.ljarka.bakingapp;

import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.widget.Toast;

import com.github.ljarka.bakingapp.detail.RecipeDetailActivity;
import com.github.ljarka.bakingapp.network.model.Recipe;

import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.disposables.Disposables;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

public class MainActivity extends AppCompatActivity implements OnRecipeItemClickListener {

    private RecipesAdapter recipesAdapter = new RecipesAdapter();
    private Disposable disposable = Disposables.empty();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        MainViewModel viewModel = ViewModelProviders.of(this).get(MainViewModel.class);
        RecyclerView recyclerView = findViewById(R.id.recyclerView);
        recyclerView.setLayoutManager(new GridLayoutManager(this, getResources().getInteger(R.integer.span_count)));
        recyclerView.setAdapter(recipesAdapter);
        recipesAdapter.setOnRecipeItemClickListener(this);
        disposable = viewModel.getRecipes()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Consumer<List<Recipe>>() {
                    @Override
                    public void accept(List<Recipe> recipes) throws Exception {
                        recipesAdapter.setRecipes(recipes);
                    }
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) throws Exception {
                        Toast.makeText(MainActivity.this, "Error", Toast.LENGTH_LONG).show();
                    }
                });
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        disposable.dispose();
    }

    @Override
    public void onRecipeItemClick(@Nullable Recipe recipe) {
        if (recipe != null) {
            startActivity(RecipeDetailActivity.createIntent(this, recipe));
        }
    }
}

File: app/src/main/java/com/github/ljarka/bakingapp/MainActivity.kt
package com.github.ljarka.bakingapp

import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.widget.Toast
import com.github.ljarka.bakingapp.detail.RecipeDetailActivity
import com.github.ljarka.bakingapp.network.model.Recipe
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity(), OnRecipeItemClickListener {
    private val adapter: RecipesAdapter = RecipesAdapter()

    private lateinit var viewModel: MainViewModel
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        viewModel = ViewModelProviders.of(this).get(MainViewModel::class.java)
        recyclerView.layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        recyclerView.adapter = adapter
        adapter.onRecipeItemClickListener = this
        viewModel.getRecipes()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    adapter.recipes = it
                }, {
                    Toast.makeText(this@MainActivity, "Error", Toast.LENGTH_LONG).show()
                })
    }

    override fun onRecipeItemClick(recipe: Recipe?) {
        if (recipe != null) {
            RecipeDetailActivity.createIntent(this, recipe)
        }
    }
}
